### PR TITLE
Add Equals and Hashcode to ApiErrorBase / ApiErrorWithMetadata

### PR DESCRIPTION
**Pull request by rabeyta**
Add equals and hashcode methods to ApiErrorBase and ApiErrorWithMetadata. Modify ApiErrorComparator to use new comparison and sort by ErrorCode after Name.

### CODE REVIEW

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java [hunk 1]
```diff
@@ -2,27 +2,34 @@
...
 
     @Test(expected = IllegalArgumentException.class)
-    public void constructorShouldThrowIllegalArgumentExceptionIfPassedNullName() {
+    public void constructor_should_throw_IllegalArgumentException_null_name() {
         new ApiErrorBase(null, 42, "some error", 400);
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void constructor_should_throw_IllegalArgumentException_null_error_code() {
+        new ApiErrorBase(null, null, "some error", 400, null);
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java]: The `IllegalArgumentException` being thrown here is likely due to the null name, not the null error code you're trying to test. I'd recommend switching from `@Test(expected = ...)` to AssertJ's `catchThrowable` and verify the message so you know the exception was thrown for the desired reason.

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorWithMetadataTest.java [hunk 2]
```diff
@@ -212,4 +212,65 @@ public void convenience_constructor_gracefully_ignores_null_pairs() {
...
+
+        String otherClass = useNull? null : "";
+
+        assertThat(awm.equals(otherClass)).isFalse();
+    }
+
+    @Test
+    @DataProvider(value = {
+            "false | false | false | false | false | true  | true  ",
+            "false | true  | false | false | false | false | false ",
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorWithMetadataTest.java]: Same missing data provider row.

File: backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Test
+    public void constructor_code_coverage() {
+        ApiErrorUtil util = new ApiErrorUtil();
+        assertThat(util).isNotNull();
+    }
+
+    @Test
+    public void generate_api_error_hashcode_generates_expected_hashcodes() {
+        Map<String, Object> metadata = MapBuilder.<String, Object>builder().put("foo", UUID.randomUUID().toString()).build();
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java]: I would duplicate the `Objects.hash(...)` here in the test and verify that the hashcode returned by the method matches the expected `Objects.hash(...)` value. It's a bit of a tautology, but works as a sanity check and can catch if the method gets updated in the future to make sure we take a step back and ask if it's a change we really want.

File: backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+        String otherClass = secondClassNull? null : "";
+
+        assertThat(isApiErrorEqual(firstClassNull? null : apiError, otherClass)).isEqualTo(expectedValue);
+    }
+
+    @Test
+    @DataProvider(value = {
+            "false | false | false | false | false | true  ",
+            "false | true  | false | false | false | false ",
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java]: Same missing data provider row.

Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java]: good eyes. fixed.

File: backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java (hunk 3 above)
Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/util/ApiErrorUtilTest.java]: good call. i've updated

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorWithMetadataTest.java (hunk 2 above)
Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorWithMetadataTest.java]: good eyes. added.

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorComparatorTest.java [hunk 5]
```diff
@@ -56,6 +56,23 @@ public void should_use_name_comparison_when_args_are_valid() {
...
     @Test
+    public void should_use_error_code_comparison_when_args_are_valid() {
+        ApiError mockApiError1 = mock(ApiError.class);
+        ApiError mockApiError2 = mock(ApiError.class);
+
+        String name = UUID.randomUUID().toString();
+        doReturn(name).when(mockApiError1).getName();
+        doReturn(name).when(mockApiError2).getName();
+
+        String errorCode1 = "1";
```
[the remark this answers is not in the document]
Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorComparatorTest.java]: I've made those suggestions.

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java [hunk 6]
```diff
@@ -31,16 +38,91 @@ public void mirrorConstructorShouldCopyAllFields() {
...
+
+        String otherClass = useNull? null : "";
+
+        assertThat(error.equals(otherClass)).isFalse();
+    }
+
+    @Test
+    @DataProvider(value = {
+            "false | false | false | false | false | true  ",
+            "false | true  | false | false | false | false ",
```
[the remark this answers is not in the document]
Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java]: good eyes.added!

File: backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java (hunk 1 above)
Comment (rabeyta) [reply backstopper-core/src/test/java/com/nike/backstopper/apierror/ApiErrorBaseTest.java]: i've changed this test and the other in the class to use `catchThrowable`

File: backstopper-core/src/main/java/com/nike/backstopper/util/ApiErrorUtil.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package com.nike.backstopper.util;
+
+import com.nike.backstopper.apierror.ApiError;
+
+import java.util.Objects;
+
+/**
+ * Utility class to enable sharing code between {@link ApiError} implementations. This probably won't be used outside
```
[the remark this answers is not in the document]
Comment (rabeyta) [reply backstopper-core/src/main/java/com/nike/backstopper/util/ApiErrorUtil.java]: updated

File: backstopper-core/src/main/java/com/nike/backstopper/apierror/ApiErrorComparator.java [hunk 8]
```diff
@@ -1,11 +1,13 @@
...
 import java.util.Comparator;
-import java.util.Objects;
+
+import static com.nike.backstopper.util.ApiErrorUtil.generateApiErrorHashCode;
+import static com.nike.backstopper.util.ApiErrorUtil.isApiErrorEqual;
 
 /**
  * A comparator that knows how to compare {@link ApiError} instances by {@link ApiError#getName()} first, and then by
- * {@link ApiError#getMetadata()} to allow for multiple same-named errors that have different metadata.
+ * {@link ApiError#getErrorCode()} ()} to allow for multiple same-named errors that have different metadata.
```
[the remark this answers is not in the document]
Comment (rabeyta) [reply backstopper-core/src/main/java/com/nike/backstopper/apierror/ApiErrorComparator.java]: updated

**Comment by nicmunroe** [comment]
Fantastic - thank you @rabeyta!

**Pull request merged**